### PR TITLE
Changing the Socket Function to a Class

### PR DESCRIPTION
**Pull request by DedShotTM**

### CODE REVIEW

File: src/gateway/WebSocket.ts [hunk 1]
```diff
@@ -1,41 +1,56 @@
+/* eslint-disable @typescript-eslint/member-ordering */
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:1]: Please don't globally disable the rule for the file

File: src/gateway/WebSocket.ts [hunk 2]
```diff
@@ -1,41 +1,56 @@
...
+  private ws!: WebSocket;
+
+  constructor(token: string, wsURL: string, intents: number) {
+    this._token = token;
+    this._wsURL = wsURL;
+    this._intents = intents;
+
+    this.identify = this.identify.bind(this);
+    this.onClose = this.onClose.bind(this);
+    this.onMessage = this.onMessage.bind(this);
```
Comment (bsian03) [inline src/gateway/WebSocket.ts:25]: Isn't necessary

**Review by bsian03 — CHANGES_REQUESTED** [review]
We should use consistent naming conventions. Using both camelCase and snake_case isn't great. @Khaaz need your thoughts on this

File: src/gateway/WebSocket.ts (hunk 2 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts:23]: it doesn't work without that, just tried

Comment (bsian03) [reply src/gateway/WebSocket.ts:23]: I hate JS and `this` bindings sometimes

**Comment by bsian03** [comment]
Fixed the CI, sorry about that

File: src/gateway/WebSocket.ts [hunk 3]
```diff
@@ -3,39 +3,52 @@ import WebSocket from 'ws';
...
+  private seq!: number;
+  private session_id!: string;
+  private ws!: WebSocket;
+
+  constructor(token: string, wsURL: string, intents: number) {
+    this._token = token;
+    this._wsURL = wsURL;
+    this._intents = intents;
+
+    this.identify = this.identify.bind(this);
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:21]: While this shouldn't be an issue at runtime, I'd probably use a different naming for the bound function? @bsian03 though

File: src/gateway/WebSocket.ts [hunk 4]
```diff
@@ -3,39 +3,52 @@ import WebSocket from 'ws';
...
-      } else lastHeartbeatAck = false;
+      if (!this.lastHeartbeatAck) {
+        clearInterval(this.interval);
+        this.ws.removeEventListener('close');
+        this.ws.terminate();
+
+        this.newWS();
+        this.resume();
+        this.lastHeartbeatAck = true;
+      } else this.lastHeartbeatAck = false;
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:36]: idk what is the configuration in the linter here, I personally prefer to always use new lines and brackets. If that's the choosen configuration we can stick with it, but if that's indifferent I feel like it makes the code clearer and spaced out...

File: src/gateway/WebSocket.ts [hunk 5]
```diff
@@ -3,39 +3,52 @@ import WebSocket from 'ws';
...
 
-export default function Socket(token: string, wsURL: string, intents: number) {
-  let interval: NodeJS.Timeout;
-  let lastHeartbeatAck = true;
-  let session_id: string;
-  let seq: number;
-  let ws: WebSocket;
-
-  function heartbeat(expected: boolean) {
+export default class Socket {
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:6]: This should at least have the same naming as the file name.
I personally would call that Client. since we are in gateway folder.
We could also go with WebSocketClient, DiscordWebSocket, etc...
It depends what we ultimately want:
- this class will emits node events for every discord events
- this class is given an instance of EventEmitter when constructed. The class emitting events will be this class.

I suggest the second solution as it seems to better separate responsabilities. 
In this case we should probably go with `DiscordWebSocket` (concrete class to handle websocket connection with discord).

File: src/gateway/WebSocket.ts [hunk 6]
```diff
@@ -3,39 +3,52 @@ import WebSocket from 'ws';
...
+    this._token = token;
+    this._wsURL = wsURL;
+    this._intents = intents;
+
+    this.identify = this.identify.bind(this);
+    this.onClose = this.onClose.bind(this);
+    this.onMessage = this.onMessage.bind(this);
+  }
+
+  heartbeat(expected: boolean) {
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:26]: all methods should be private at least while we don't have an explicit usage outside of the class. Which we shouldn't need think.

Comment (Khaaz) [reply src/gateway/WebSocket.ts:26]: The only thing we should expose is something like a `connect` or `start` method for initialising the ws connection.

File: src/gateway/WebSocket.ts [hunk 7]
```diff
@@ -45,73 +58,72 @@ export default function Socket(token: string, wsURL: string, intents: number) {
...
       case GATEWAY_OPCODES.HEARTBEAT:
-        heartbeat(false);
+        this.heartbeat(false);
         break;
 
       case GATEWAY_OPCODES.RECONNECT:
-        clearInterval(interval);
-        ws.removeEventListener('close');
-        ws.terminate();
+        clearInterval(this.interval);
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:90]: this should probably be in it's own `reset` or `clear` method.

File: src/gateway/WebSocket.ts [hunk 8]
```diff
@@ -45,73 +58,72 @@ export default function Socket(token: string, wsURL: string, intents: number) {
...
-      ws.send(JSON.stringify({
-        op: GATEWAY_OPCODES.RESUME,
-        d: {
-          token,
-          session_id,
-          seq,
-        },
-      }));
-    }
+  newWS() {
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:61]: this could be named `create` since or `initialise` instead...

File: src/gateway/WebSocket.ts [hunk 9]
```diff
@@ -45,73 +58,72 @@ export default function Socket(token: string, wsURL: string, intents: number) {
...
       case GATEWAY_OPCODES.INVALID_SESSION:
         setTimeout(() => {
-          identify();
+          this.identify();
         }, 4000);
         break;
 
       case GATEWAY_OPCODES.HELLO:
-        interval = setInterval(() => heartbeat(true), d.heartbeat_interval);
+        this.interval = setInterval(() => this.heartbeat(true), d.heartbeat_interval);
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:105]: this should be named `heartBeatInterval` instead of only `interval`

Comment (Khaaz) [reply src/gateway/WebSocket.ts:105]: Is it possible to receive two `HELLO` without having cleared the interval?

File: src/gateway/WebSocket.ts [hunk 10]
```diff
@@ -45,73 +58,72 @@ export default function Socket(token: string, wsURL: string, intents: number) {
...
-        heartbeat(false);
+        this.heartbeat(false);
         break;
 
       case GATEWAY_OPCODES.RECONNECT:
-        clearInterval(interval);
-        ws.removeEventListener('close');
-        ws.terminate();
+        clearInterval(this.interval);
+        this.ws.removeEventListener('close');
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:91]: I checked ws code, and it extends directly EventEmitter. You should use `off` or `removeListener` instead of `removeEventListener`. I would probably use `off` since you use `on` and not `addListener`.

File: src/gateway/WebSocket.ts [hunk 11]
```diff
@@ -45,73 +58,72 @@ export default function Socket(token: string, wsURL: string, intents: number) {
...
-  }
-
-  function newWS() {
-    ws = new WebSocket(wsURL);
-    ws.on('open', identify);
-    ws.on('close', onClose);
-    ws.on('message', onMessage);
+  resume() {
+    if (this.ws.readyState === WebSocket.OPEN) {
+      this.ws.send(JSON.stringify({
```
Comment (Khaaz) [inline src/gateway/WebSocket.ts:119]: I would abstract the `send` method in its own method for the sake of it. Not 100% necessary but it can make sense.
This function would just do the send and the JSON.stringify of the packet you give in param.
Potentially we could try to make it a bit fancier if there is stuff we always need to send (like token, session_id and seq but even those are not needed in the heartbeat command for instance)

**Review by Khaaz — CHANGES_REQUESTED** [review]
To sum up:
some nitpick, some request for naming changes or convention and one of two things we might want to look in more details.
As for the naming convention, as I mentionned in Discord. I would go for camelCase for all our API. And keep discord casing for everything/every data we receive or send to discord.

**Comment by Khaaz** [comment]
> We should use consistent naming conventions. Using both camelCase and snake_case isn't great. @Khaaz need your thoughts on this

If that is about something like `session_id` I would use camelCase for this one since it's used internally only. The only thing we would keep in discord casing would be:
- the data we send to discord in REST (whatever the user provide to rest functions)
- the data we receive from discord in GW (whatever we emit in gw and the user listen to)

File: src/gateway/WebSocket.ts (hunk 9 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts:105]: I don't think so, the only way to receive `HELLO` is to start a new connection, which won't happen before the last connection is terminated

File: src/gateway/WebSocket.ts (hunk 5 above)
Comment (DedShotTM) [reply src/gateway/WebSocket.ts]: we'll probably do the EventEmitter part in the next pull request
changed the names btw, used DiscordWebSocket

File: src/gateway/WebSocketClient.ts [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  initialize() {
+    this.ws = new WebSocket(this._wsURL);
+    this.ws.on('open', this.identify);
+    this.ws.on('close', this.onClose);
+    this.ws.on('message', this.onMessage);
+  }
+
+  private heartbeat(expected: boolean) {
```
Comment (bsian03) [inline src/gateway/WebSocketClient.ts:33]: We should prefix these with `_` even though they're marked private

File: src/gateway/WebSocketClient.ts [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  private seq!: number;
+  private session_id!: string;
+  private ws!: WebSocket;
+
+  constructor(token: string, wsURL: string, intents: number) {
+    this._token = token;
+    this._wsURL = wsURL;
+    this._intents = intents;
+
+    this.identify = this.identify.bind(this);
```
Comment (bsian03) [inline src/gateway/WebSocketClient.ts:21]: Original review https://github.com/libraryofcode/dis.ts/pull/15#discussion_r548976462

Comment (bsian03) [reply src/gateway/WebSocketClient.ts:21]: I think it's fine as it is

File: src/gateway/WebSocketClient.ts [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+  private heartbeat(expected: boolean) {
+    if (expected) {
+      if (!this.lastHeartbeatAck) {
+        this.terminate();
+        this.initialize();
+        this.resume();
+        this.lastHeartbeatAck = true;
+      } else {
+        this.lastHeartbeatAck = false;
+      }
```
Comment (bsian03) [inline src/gateway/WebSocketClient.ts:42]: Original review https://github.com/libraryofcode/dis.ts/pull/15#discussion_r548976664

Comment (bsian03) [reply src/gateway/WebSocketClient.ts:42]: I can update linting rules at some point in the future to prevent this. Someone please remind me

File: src/gateway/WebSocketClient.ts (hunk 12 above)
Comment (bsian03) [reply src/gateway/WebSocketClient.ts:33]: Original reviews https://github.com/libraryofcode/dis.ts/pull/15#discussion_r548977896 https://github.com/libraryofcode/dis.ts/pull/15#discussion_r548978018

Comment (bsian03) [reply src/gateway/WebSocketClient.ts:33]: We have `initialize` for that but I don't know if that's a great name, also `initialize` will currently connect to the gateway (ref https://github.com/libraryofcode/dis.ts/pull/15#discussion_r548978209)

**Review by bsian03 — COMMENTED** [review]
These are mainly just followup comments
I'd be more inclined to using snake_case for `sessionID` given the context and usage

File: src/gateway/WebSocketClient.ts (hunk 13 above)
Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts:21]: yeah we'll keep it like that

File: src/gateway/WebSocketClient.ts (hunk 12 above)
Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts:33]: > 
> 
> We should prefix these with `_` even though they're marked private

You want to prefix all private functions with `_`?

Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts:33]: > 
> 
> We have `initialize` for that but I don't know if that's a great name, also `initialize` will currently connect to the gateway (ref [#15 (comment)](https://github.com/libraryofcode/dis.ts/pull/15#discussion_r548978209))

yeah it will, what name do you suggest?

Comment (bsian03) [reply src/gateway/WebSocketClient.ts:33]: > You want to prefix all private functions with `_`?

yes please

Comment (bsian03) [reply src/gateway/WebSocketClient.ts:33]: > yeah it will, what name do you suggest?
`connect` or `start` as Khaaz mentioned

Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts]: alright both done, we'll use connect()

File: src/gateway/WebSocketClient.ts [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+  private _resume() {
+    this._sendWS(GATEWAY_OPCODES.RESUME, {
+      token: this._token,
+      sessionID: this.sessionID,
+      seq: this.seq,
+    });
+  }
+
+  private _sendWS(op: GATEWAY_OPCODES, data: any) {
```
Comment (bsian03) [inline src/gateway/WebSocketClient.ts:113]: @Khaaz Should we mark this public?

Comment (Khaaz) [reply src/gateway/WebSocketClient.ts:113]: We could. Exposing a method to manually send WS stuff is a good idea.

The plan is to have a GatewayClient class that owns a WebsocketClient class. GatewayClient will emits events when WebsocketClient receive events from discord. It will also send ws stuff so sendWS needs to be accessible from outside.
TLDR: it should indeed be public. That said it should also be named `send` only. we are already in WS class, no need to specify it again in the method name.

File: src/gateway/WebSocketClient.ts [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+  private _onMessage(data: Payload) {
+    data = JSON.parse(String(data));
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { t, s, op, d } = data;
+    if (s) this.seq = s;
+
+    switch (op) {
+      case GATEWAY_OPCODES.DISPATCH:
+        // if (EVENTS[t]) client.emit(EVENTS[t], d);
+        this.sessionID = d.sessionID;
```
Comment (Khaaz) [inline src/gateway/WebSocketClient.ts:73]:
```suggestion
        this.sessionID = d.session_id;
```
Mistake when renaming

File: src/gateway/WebSocketClient.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+      default:
+        break;
+    }
+  }
+
+  private _resume() {
+    this._sendWS(GATEWAY_OPCODES.RESUME, {
+      token: this._token,
+      sessionID: this.sessionID,
```
Comment (Khaaz) [inline src/gateway/WebSocketClient.ts:108]:
```suggestion
      session_id: this.sessionID,
```

File: src/gateway/WebSocketClient.ts [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+      case GATEWAY_OPCODES.RECONNECT:
+        this._terminate();
+        this.connect();
+        this._resume();
+        break;
+
+      case GATEWAY_OPCODES.INVALID_SESSION:
+        setTimeout(() => {
+          this._identify();
+        }, 4000);
```
Comment (Khaaz) [inline src/gateway/WebSocketClient.ts:89]: We might want this as a constant. At least as a local constant in this file

File: src/gateway/WebSocketClient.ts [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+  private _wsURL: string;
+  private heartBeatInterval!: NodeJS.Timeout;
+  private lastHeartbeatAck = true;
+  private seq!: number;
+  private sessionID!: string;
+  private ws!: WebSocket;
+
+  constructor(token: string, wsURL: string, intents: number) {
+    this._token = token;
+    this._wsURL = wsURL;
```
Comment (Khaaz) [inline src/gateway/WebSocketClient.ts:18]: same as sendWS, could simply be named `url`? Less important than the other one as not exposed though

**Review by Khaaz — CHANGES_REQUESTED** [review]
Almost there, few changes and naming. Also `session_id` you missed when renaming.
Idk what we agreed to in the end, but to come back to this class naming, I think it should actually be named `DiscordWebSocket`. This makes the more sense imho.

File: src/gateway/WebSocketClient.ts (hunk 18 above)
Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts:98]: how should we call the constant

File: src/gateway/WebSocketClient.ts (hunk 17 above)
Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts]: nice catch

File: src/gateway/WebSocketClient.ts (hunk 18 above)
Comment (bsian03) [reply src/gateway/WebSocketClient.ts:98]:
```ts
class Name {
  public IDENTIFY_TIMEOUT = 4000
  // ...
  setTimeout(() => {
    this._identify();
  }, this.IDENTIFY_TIMEOUT);
  // ...
}
``` something along the lines of this

File: src/gateway/WebSocketClient.ts [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+  private _identify() {
+    this.send(GATEWAY_OPCODES.IDENTIFY, {
+      token: this._token,
+      intents: this._intents,
+      properties: {
+        $os: process.platform,
+        $browser: 'Dis.ts',
+        $device: 'Dis.ts',
+      },
```
Comment (bsian03) [inline src/gateway/WebSocketClient.ts:64]: At some point we might want to allow people to input whatever they want here.. that's for a later time though

File: src/gateway/WebSocketClient.ts (hunk 18 above)
Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts]: done

File: src/gateway/WebSocketClient.ts [hunk 21]
```diff
@@ -0,0 +1,128 @@
+import WebSocket from 'ws';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { EVENTS, Payload } from './constants';
+import { GATEWAY_OPCODES, GATEWAY_CLOSE_EVENT_CODES } from '../util/Constants';
+
+export default class WebSocketClient {
+  public IDENTIFY_TIMEOUT = 4000;
```
Comment (Khaaz) [inline src/gateway/WebSocketClient.ts:7]: any reason this is public?
I would honestly have this as a constant outside of the class entirely (like right above the class)
At the very least, private and readonly.

Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts]: done

File: src/gateway/WebSocketClient.ts [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { EVENTS, Payload } from './constants';
+import { GATEWAY_OPCODES, GATEWAY_CLOSE_EVENT_CODES } from '../util/Constants';
+const IDENTIFY_TIMEOUT = 4000;
+
+export default class WebSocketClient {
+  private _intents: number;
+  private _token: string;
+  private _url: string;
+  private heartBeatInterval!: NodeJS.Timeout;
```
Comment (Khaaz) [inline src/gateway/WebSocketClient.ts:11]: Any reasons those properties are not prefixed?

**Review by Khaaz — COMMENTED** [review]
Last nitpick, other than that, looks good to me.

File: src/gateway/WebSocketClient.ts (hunk 22 above)
Comment (DedShotTM) [reply src/gateway/WebSocketClient.ts]: changed

**Review by Khaaz — APPROVED** [review]

File: src/gateway/WebSocketClient.ts [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+
+  send(op: GATEWAY_OPCODES, data: any) {
+    if (this._ws.readyState === WebSocket.OPEN) {
+      this._ws.send(JSON.stringify({
+        op,
+        d: data,
+      }));
+    }
+  }
+
```
Comment (bsian03) [inline src/gateway/WebSocketClient.ts:42]: Should we add some indication to show whether the payload was actually sent or not? Probably via a bool return type

Comment (Khaaz) [reply src/gateway/WebSocketClient.ts:42]: I think we can add that to the todos/.
Later all send/ will need to be queued. To handle stuff like identify rl or delay when reconnecting.
Waiting for your approval for the whole PR. If Deadshot wants to implement a boolean return though that's something we will use in the future

**Review by bsian03 — APPROVED** [review]
